File: ride-sharing-system/java-without-multithreading/strategy/MostVacantStrategy.java
package strategy;
import entities.Ride;
import java.util.List;
import java.util.Optional;

public class MostVacantStrategy implements RideSelectionStrategy {
    @Override
    public Optional<Ride> selectRide(List<Ride> availableRides) {
        Ride mostVacant = null;
        int maxSeats = -1;

        for (Ride ride : availableRides) {
            if (ride.getAvailableSeats() > maxSeats) {
                maxSeats = ride.getAvailableSeats();
                mostVacant = ride;
            }
        }
        
        return Optional.ofNullable(mostVacant);
    }
}

File: ride-sharing-system/go-without-multithreading/strategy/most_vacant_strategy.go
package strategy
import "app/ride-sharing-system/go-without-multithreading/entities"

// MostVacantStrategy calculates maximum remaining bounds natively ensuring systems minimize booking failures.
type MostVacantStrategy struct{}

func (s *MostVacantStrategy) SelectRide(availableRides []*entities.Ride) *entities.Ride {
	var mostVacant *entities.Ride
	maxSeats := -1

	// Explicitly map natively looping bounds securing max values mathematically.
	for _, ride := range availableRides {
		if ride.AvailableSeats > maxSeats {
			maxSeats = ride.AvailableSeats
			mostVacant = ride
		}
	}
	return mostVacant
}
